docs/man/man3/PowerManage.3
.TH "PowerManage" 3 "Tue May 17 2022" "EPS MCU" \" -*- nroff -*-
.ad l
.nh
.SH NAME
PowerManage \- Aktivering og deaktivering af powerrails på baggrund af OCE og state\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <PowerManage\&.h>\fP
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBModules\fP { \fBSUB_1\fP }"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static string \fBToString\fP (\fBModules\fP module)"
.br
.RI "Omdanner modul til string med navnet for modulet\&. "
.ti -1c
.RI "static void \fBOceFromIsr\fP (\fBModules\fP module)"
.br
.RI "Høj prioritetet deaktivering af en power rail ved OCE\&. "
.ti -1c
.RI "static void \fBSetOverride\fP (\fBModules\fP module, bool is_on)"
.br
.RI "Opdaterer et modul i module_override_ listen\&. Og Kalder \fBWriteSubState()\fP\&. "
.ti -1c
.RI "static void \fBUpdateEnabled\fP ()"
.br
.RI "Henter den nuværende state fra FSM biblioteket og opdaterer enabled_modules_ listen efter denne\&. Kalder til sidst \fBWriteSubState()\fP\&. "
.ti -1c
.RI "static void \fBWriteSubState\fP ()"
.br
.RI "Opdaterer den endelige tisltand af Power Rails ved at AND'e enabled_modules_ og module_override_\&. "
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static bool \fBenabled_modules_\fP [SUBMODULES_NUM] = {true}"
.br
.ti -1c
.RI "static bool \fBmodule_override_\fP [SUBMODULES_NUM] = {true}"
.br
.ti -1c
.RI "static DigitalOut \fBpower_rails_\fP [1]"
.br
.in -1c
.SH "Detailed Description"
.PP 
Aktivering og deaktivering af powerrails på baggrund af OCE og state\&. 
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBPowerManage::Modules\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fISUB_1 \fP\fP
.SH "Member Function Documentation"
.PP 
.SS "void PowerManage::OceFromIsr (\fBPowerManage::Modules\fP module)\fC [static]\fP"

.PP
Høj prioritetet deaktivering af en power rail ved OCE\&. 
.PP
\fBParameters\fP
.RS 4
\fImodule\fP Modulet der blev målt OCE på 
.RE
.PP

.SS "void PowerManage::SetOverride (\fBPowerManage::Modules\fP module, bool is_on)\fC [static]\fP"

.PP
Opdaterer et modul i module_override_ listen\&. Og Kalder \fBWriteSubState()\fP\&. 
.PP
\fBParameters\fP
.RS 4
\fImodule\fP Det modul der skal aktiveres/deaktiveres 
.br
\fIis_on\fP false = deaktiver | true = aktiver 
.RE
.PP

.SS "string PowerManage::ToString (\fBPowerManage::Modules\fP module)\fC [static]\fP"

.PP
Omdanner modul til string med navnet for modulet\&. 
.PP
\fBParameters\fP
.RS 4
\fImodule\fP Det modul der ønskes navn på 
.RE
.PP
\fBReturns\fP
.RS 4
string med navnet for modulet 
.RE
.PP

.SS "void PowerManage::UpdateEnabled ()\fC [static]\fP"

.PP
Henter den nuværende state fra FSM biblioteket og opdaterer enabled_modules_ listen efter denne\&. Kalder til sidst \fBWriteSubState()\fP\&. 
.SS "void PowerManage::WriteSubState ()\fC [static]\fP"

.PP
Opdaterer den endelige tisltand af Power Rails ved at AND'e enabled_modules_ og module_override_\&. 
.SH "Member Data Documentation"
.PP 
.SS "bool PowerManage::enabled_modules_ = {true}\fC [static]\fP, \fC [private]\fP"

.SS "bool PowerManage::module_override_ = {true}\fC [static]\fP, \fC [private]\fP"

.SS "DigitalOut PowerManage::power_rails_\fC [static]\fP, \fC [private]\fP"
\fBInitial value:\fP
.PP
.nf
= {
    RAIL_1_PIN
}
.fi


.SH "Author"
.PP 
Generated automatically by Doxygen for EPS MCU from the source code\&.


lib/PowerManage/PowerManage.cpp
//
// Created by EIT-416 on 5/2/22.
//

#include "Fsm.h"
#include "PowerManage.h"

bool PowerManage::enabled_modules_[SUBMODULES_NUM] = {true};
bool PowerManage::module_override_[SUBMODULES_NUM] = {true};
DigitalOut PowerManage::power_rails_[] = {
    RAIL_1_PIN
};

/**
 * @brief Henter den nuværende state fra FSM biblioteket og
 *        opdaterer enabled_modules_ listen efter denne. Kalder
 *        til sidst WriteSubState().
 */

void PowerManage::UpdateEnabled() {

  Fsm::State current_state = Fsm::GetCurrentState();

  switch (current_state) {

    case Fsm::EMERGENCY:enabled_modules_[0] = false;
//            enabled_modules_[1] = false;
//            enabled_modules_[2] = false;
//            enabled_modules_[3] = false;
//            enabled_modules_[4] = false;
      break;
    case Fsm::LOW:enabled_modules_[0] = true;
//            enabled_modules_[1] = false;
//            enabled_modules_[2] = false;
//            enabled_modules_[3] = false;
//            enabled_modules_[4] = false;
      break;
    case Fsm::MID_LOW:enabled_modules_[0] = true;
//            enabled_modules_[1] = true;
//            enabled_modules_[2] = false;
//            enabled_modules_[3] = false;
//            enabled_modules_[4] = false;
      break;
    case Fsm::MID:enabled_modules_[0] = true;
//            enabled_modules_[1] = true;
//            enabled_modules_[2] = true;
//            enabled_modules_[3] = false;
//            enabled_modules_[4] = false;
      break;
    case Fsm::MID_HIGH:enabled_modules_[0] = true;
//            enabled_modules_[1] = true;
//            enabled_modules_[2] = true;
//            enabled_modules_[3] = true;
//            enabled_modules_[4] = false;
      break;
    case Fsm::HIGH:enabled_modules_[0] = true;
//            enabled_modules_[1] = true;
//            enabled_modules_[2] = true;
//            enabled_modules_[3] = true;
//            enabled_modules_[4] = true;

      break;
  }
  WriteSubState();
}

/**
 * @brief Opdaterer et modul i module_override_ listen. Og Kalder WriteSubState().
 * 
 * @param module Det modul der skal aktiveres/deaktiveres
 * @param is_on  false = deaktiver | true = aktiver
 */

void PowerManage::SetOverride(PowerManage::Modules module, bool is_on) {
  switch (module) {
    case SUB_1:module_override_[0] = is_on;
      break;
//    case SUB_2:
//            module_override_[1] = is_on;
//      break;
//    case SUB_3:
//            module_override_[2] = is_on;
//      break;
//    case SUB_4:
//            module_override_[3] = is_on;
//      break;
//    case SUB_5:
//            module_override_[4] = is_on;
      break;
  }
  WriteSubState();
}

/**
 * @brief Omdanner modul til string med navnet for modulet
 * 
 * @param module Det modul der ønskes navn på
 * @return string med navnet for modulet
 */
string PowerManage::ToString(PowerManage::Modules module) {
  switch (module) {
    case SUB_1:return "Submodule 1";
    default: return "Unknown";
  }
}

/**
 * @brief Opdaterer den endelige tisltand af Power Rails ved at AND'e
 *        enabled_modules_ og module_override_
 * 
 */

void PowerManage::WriteSubState() {
  for (int i = 0; i < SUBMODULES_NUM; i++) {
    power_rails_[i].write(enabled_modules_[i] and module_override_[i]);
  }
}

/**
 * @brief Høj prioritetet deaktivering af en power rail ved OCE
 * 
 * @param module Modulet der blev målt OCE på
 */

void PowerManage::OceFromIsr(PowerManage::Modules module) {
  power_rails_[module].write(false);
}
